streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorNode.java

```diff
@@ -214,9 +214,13 @@ public void process(final Record<KIn, VIn> record) {
                 internalProcessorContext.currentNode().name(),
                 internalProcessorContext.taskId());
 
-            final ProcessingExceptionHandler.ProcessingHandlerResponse response = processingExceptionHandler
-                .handle(errorHandlerContext, record, e);
+            final ProcessingExceptionHandler.ProcessingHandlerResponse response;
 
+            try {
+                response = processingExceptionHandler.handle(errorHandlerContext, record, e);
+            } catch (final Exception fatalUserException) {
+                throw new StreamsException("Fatal user code error in processing error callback", fatalUserException);
```

Comment:
Should we add an error log before throwing?

Btw: it seems we have such a guard only for the `DeserializationExceptionHandler#handle(...)` call, but not for the `ProductionExceptionHanlder#call(...)`? Might be good to do a follow up PR to add it to the proction-error-handler case, too?

Reply:
One more thing: if we throw `StreamsException` here, we might catch it in an upstream `process()`, right? So should we rather wrap it with a `FailedProcessingException`?

Reply:
done: 
I changed the StreamException into a FailedProcessingException
added some test to check that no more processor is executed when an exception occurs or when an exception occurs in a handler.

Reply:
> Should we add an error log before throwing?
> 
> Btw: it seems we have such a guard only for the `DeserializationExceptionHandler#handle(...)` call, but not for the `ProductionExceptionHanlder#call(...)`? Might be good to do a follow up PR to add it to the proction-error-handler case, too?

@mjsax is it mandatory for this KIP ?

Reply:
> One more thing: if we throw `StreamsException` here, we might catch it in an upstream `process()`, right? So should we rather wrap this `StreamsException` with a `FailedProcessingException`?

Correct. `FailedProcessingException` is thrown as is:

https://github.com/apache/kafka/blob/683b8a2beeb5a75ef0a2ac878cf5a963f9b9d915/streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorNode.java#L205

> Do we actually have a test for "passing through" of `FailedProcessingException`

We do:

https://github.com/apache/kafka/blob/683b8a2beeb5a75ef0a2ac878cf5a963f9b9d915/streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java#L130

It asserts `FailedProcessingException` is not handled and thrown as is. Another test has been added to ensure that the cause of the caught exception is from the processing exception handler.

Reply:
>> Btw: it seems we have such a guard only for the DeserializationExceptionHandler#handle(...) call, but not for the ProductionExceptionHanlder#call(...)? Might be good to do a follow up PR to add it to the proction-error-handler case, too?
>
> @mjsax is it mandatory for this KIP ?

Well, we don't want a handler which throws an exception to **not** crash KS, because the exception is caught in an upstream handler, right? But should be a simple thing to add?

Reply:
> > > Btw: it seems we have such a guard only for the DeserializationExceptionHandler#handle(...) call, but not for the ProductionExceptionHanlder#call(...)? Might be good to do a follow up PR to add it to the proction-error-handler case, too?
> > 
> > 
> > @mjsax is it mandatory for this KIP ?
> 
> Well, we don't want a handler which throws an exception to **not** crash KS, because the exception is caught in an upstream handler, right? But should be a simple thing to add?
> 
> But I see your point that it's not really related to the KIP. I can do a follow up myself.

Ok, I will open a new PR for that !